### PR TITLE
Tuple based session key module.

### PR DESCRIPTION
**Pull request by dvc94ch**
The session module already supports tuple based session keys. You can add a custom session key type and provide `Into<sr25519::Public>` and `Into<ed25519::Public>` or `Into<ble::Public>` implementations. I updated the test to include such an example. Please let me know if I've missed the point :) @gavofyork @Kianenigma

### CODE REVIEW

**Comment by rphmeier** [comment]
I'm fine with the "just have your `SessionKey` type be a tuple" approach -- but we will need a little further glue for consensus engines to adapt `SessionKey` into their consensus key type.

e.g. we have two keys `(a, b)` for two consensus engines `(A, B)`. We probably want something like 

```rust
trait MyConsensusTrait: session::Trait {
    type OurKey;
    type ConvertSessionKey: Convert<<Self as session::Trait>::SessionKey, Self::OurKey>;
}
```

**Comment by dvc94ch** [comment]
So this glue has to go into the aura::Trait, grandpa::Trait and babe::Trait?

File: srml/session/src/lib.rs [hunk 1]
```diff
@@ -356,9 +358,40 @@ mod tests {
...
 		type Log = DigestItem;
-		type SessionKey = UintAuthorityId;
+		type SessionKey = TestSessionKey;
 		type InherentOfflineReport = ();
 	}
+
+	#[derive(Clone, Debug, Default, PartialEq, Eq, Encode, Decode, Serialize, Deserialize)]
+	pub struct TestSessionKey(u64, u64);
+
+	impl Into<sr25519::Public> for TestSessionKey {
```
Comment (tomusdrw) [inline srml/session/src/lib.rs]: `From<TestSessionKey> for sr25519::Public` should be preferred over `Into` (you will get `Into` for free anyway)

File: srml/session/src/lib.rs [hunk 2]
```diff
@@ -356,9 +358,40 @@ mod tests {
...
+		fn into(self) -> sr25519::Public {
+				let bytes: [u8; 32] = U256::from(self.0).into();
```
Comment (tomusdrw) [inline srml/session/src/lib.rs]: Why go through `U256` and not `H256`? Won't you end up with possible endianess issues?

Comment (dvc94ch) [reply srml/session/src/lib.rs]: I copied that from `testing.rs`. Changed it there too.

**Comment by gavofyork** [comment]
> So this glue has to go into the aura::Trait, grandpa::Trait and babe::Trait?

Yup

**Comment by burdges** [comment]
Just some considerations about BLS keys:  

It's fine if GRANDPA reads them in using some trait like `Into<bls::Public>`, but we actually want *all* validator BLS keys presented simultaneously as some type satisfying the [`bls-pixel::pop::ProofsOfPossession`](https://github.com/w3f/bls/blob/master/src/pop.rs#L167) trait, which permits public key lookups by index or public key.  Just fyi, you verify by making a [`bls-pixel::pop::BitPoPSignedMessage`](https://github.com/w3f/bls/blob/master/src/pop.rs#L270) and adding all the signatures.

Any type that satisfies the `ProofsOfPossession` trait should check the proofs-of-possession for all keys.  You can check many/all simultaneously using [`bls-pixel::distinct::DistinctMessages`](https://github.com/w3f/bls/blob/master/src/distinct.rs).  You'll might batch verify the whole session certificate at the same time.

The session keys database should ideally be pulled out from the chain's data into some in-memory database like once per validator set election.  It'll get run very slow if you end up checking the proofs-of-possession with every block or something.

**Comment by gavofyork** [comment]
@dvc94ch The original design for this probably won't work; session module will need to keep an eye on the individual keys to ensure that validators do not use any key that some other validator has registered.

Something like #2628 is therefore needed, though it should be made to work with all elements of the tuple.

So: session module itself will need to know how to deconstruct session key "tuples" into individual keys, and it should maintain a per-key *ordered* list of active session keys in order to have efficient means of prevent a duplicate being registered.

To detect duplicates

**Comment by gavofyork** [comment]
I'll write a bit of pseudocode to illustrate...

**Comment by burdges** [comment]
Some considerations about the structure listed in https://github.com/paritytech/polkadot/issues/47#issuecomment-499049810

**Pull request closed without merging**